File: .github/scripts/test-path-safe-consumption.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "find"
require "open3"
require "pathname"
require "rbconfig"
require "tmpdir"

SOURCE_ROOT = File.expand_path("../..", __dir__)
NON_MUTATING_SMOKE = ".github/scripts/test-non-mutating-consumption.rb"
SMOKES = {
  ".github/scripts/test-minimal-profile-layouts.rb" => "Minimal profile repository layout tests passed.",
  ".github/scripts/test-copyable-template-consumption.rb" => "Copyable template adoption and installation tests passed.",
  ".github/scripts/test-concrete-skill-completion.rb" => "Concrete skill completion hygiene tests passed.",
  ".github/scripts/test-parent-owned-vendoring.rb" => "Parent-owned vendoring smoke tests passed.",
  NON_MUTATING_SMOKE => "Non-mutating skill consumption smoke tests passed."
}.freeze

failures = []

def tree_snapshot(root)
  root_path = Pathname.new(root)
  Find.find(root).sort.each_with_object({}) do |path, snapshot|
    relative = path == root ? "." : Pathname.new(path).relative_path_from(root_path).to_s
    stat = File.lstat(path)
    type = if stat.directory?
             "directory"
           elsif stat.file?
             "file"
           elsif stat.symlink?
             "symlink"
           else
             "other"
           end
    record = [type, stat.mode & 0o7777, stat.mtime.to_r]
    record << File.binread(path) if stat.file?
    record << File.readlink(path) if stat.symlink?
    snapshot[relative] = record
  end
end

Dir.mktmpdir("path-safe-consumption-host") do |host_root|
  path_safe_root = File.join(host_root, "workspace with spaces", "日本語")
  poison_root = File.join(host_root, "caller-owned-git-context")
  poison_git_dir = File.join(poison_root, "git-dir")
  poison_work_tree = File.join(poison_root, "work-tree")
  poison_index = File.join(poison_root, "index")
  FileUtils.mkdir_p([path_safe_root, poison_git_dir, poison_work_tree])
  File.binwrite(poison_index, "caller-owned index sentinel\n")
  File.binwrite(File.join(poison_git_dir, "sentinel"), "git-dir sentinel\n")
  File.binwrite(File.join(poison_work_tree, "sentinel"), "work-tree sentinel\n")
  poison_before = tree_snapshot(poison_root)

  expected_root = File.realpath(path_safe_root)
  environment = {
    "TMPDIR" => path_safe_root,
    "TMP" => path_safe_root,
    "TEMP" => path_safe_root,
    "RUBYOPT" => nil
  }

  stdout, stderr, status = Open3.capture3(
    environment,
    RbConfig.ruby,
    "-rtmpdir",
    "-e",
    "print File.realpath(Dir.tmpdir)",
    chdir: SOURCE_ROOT
  )
  unless status.success? && stderr.empty? && stdout == expected_root
    failures << "temporary-root selection failed: status=#{status.exitstatus.inspect}, " \
                "stdout=#{stdout.inspect}, stderr=#{stderr.inspect}, expected=#{expected_root.inspect}"
  end

  SMOKES.each do |relative_script, success_line|
    smoke_environment = environment
    if relative_script == NON_MUTATING_SMOKE
      smoke_environment = environment.merge(
        "GIT_DIR" => poison_git_dir,
        "GIT_INDEX_FILE" => poison_index,
        "GIT_WORK_TREE" => poison_work_tree
      )
    end

    stdout, stderr, status = Open3.capture3(
      smoke_environment,
      RbConfig.ruby,
      relative_script,
      chdir: SOURCE_ROOT
    )
    unless status.success? && stderr.empty? && stdout.lines.last&.strip == success_line
      failures << "#{relative_script} failed under path-safe temporary root: " \
                  "status=#{status.exitstatus.inspect}, stdout=#{stdout.inspect}, stderr=#{stderr.inspect}"
    end

    if relative_script == NON_MUTATING_SMOKE && tree_snapshot(poison_root) != poison_before
      failures << "#{relative_script} mutated inherited caller-owned Git context"
    end
  end
end

unless failures.empty?
  failures.each { |failure| warn failure }
  exit 1
end

puts "Path-safe core consumption smoke tests passed."

File: .github/scripts/test_path_safe_consumption.py
#!/usr/bin/env python3
"""Rerun portable core smokes beneath paths with spaces and non-ASCII text."""

from __future__ import annotations

import os
import stat
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Any

SOURCE_ROOT = Path(__file__).resolve().parents[2]
NON_MUTATING_SMOKE = ".github/scripts/test_non_mutating_consumption.py"
SMOKES = {
    ".github/scripts/test_minimal_profile_layouts.py": "Minimal profile repository layout tests passed.",
    ".github/scripts/test_copyable_template_consumption.py": "Copyable template adoption and installation tests passed.",
    ".github/scripts/test_concrete_skill_completion.py": "Concrete skill completion hygiene tests passed.",
    ".github/scripts/test_parent_owned_vendoring.py": "Parent-owned vendoring smoke tests passed.",
    NON_MUTATING_SMOKE: "Non-mutating skill consumption smoke tests passed.",
}


def snapshot_record(path: Path) -> tuple[Any, ...]:
    info = path.lstat()
    mode = stat.S_IMODE(info.st_mode) | (info.st_mode & (stat.S_ISUID | stat.S_ISGID | stat.S_ISVTX))
    if stat.S_ISDIR(info.st_mode):
        kind = "directory"
    elif stat.S_ISREG(info.st_mode):
        kind = "file"
    elif stat.S_ISLNK(info.st_mode):
        kind = "symlink"
    else:
        kind = "other"
    record: list[Any] = [kind, mode & 0o7777, info.st_mtime_ns]
    if kind == "file":
        record.append(path.read_bytes())
    elif kind == "symlink":
        record.append(os.readlink(path))
    return tuple(record)


def tree_snapshot(root: Path) -> dict[str, tuple[Any, ...]]:
    snapshot: dict[str, tuple[Any, ...]] = {".": snapshot_record(root)}

    def visit(directory: Path) -> None:
        with os.scandir(directory) as iterator:
            entries = sorted(iterator, key=lambda entry: entry.name)
        for entry in entries:
            path = Path(entry.path)
            relative = path.relative_to(root).as_posix()
            snapshot[relative] = snapshot_record(path)
            if entry.is_dir(follow_symlinks=False):
                visit(path)

    visit(root)
    return snapshot


def run(
    command: list[str], *, cwd: Path, environment: dict[str, str]
) -> subprocess.CompletedProcess[str]:
    env = os.environ.copy()
    env.update(environment)
    env.pop("PYTHONPATH", None)
    return subprocess.run(
        command,
        cwd=cwd,
        env=env,
        text=True,
        encoding="utf-8",
        capture_output=True,
        check=False,
    )


def main() -> int:
    failures: list[str] = []
    with tempfile.TemporaryDirectory(prefix="path-safe-consumption-host") as temporary:
        host_root = Path(temporary)
        path_safe_root = host_root / "workspace with spaces/日本語"
        poison_root = host_root / "caller-owned-git-context"
        poison_git_dir = poison_root / "git-dir"
        poison_work_tree = poison_root / "work-tree"
        poison_index = poison_root / "index"
        for path in (path_safe_root, poison_git_dir, poison_work_tree):
            path.mkdir(parents=True, exist_ok=True)
        poison_index.write_bytes(b"caller-owned index sentinel\n")
        (poison_git_dir / "sentinel").write_bytes(b"git-dir sentinel\n")
        (poison_work_tree / "sentinel").write_bytes(b"work-tree sentinel\n")
        poison_before = tree_snapshot(poison_root)

        expected_root = str(path_safe_root.resolve())
        environment = {
            "TMPDIR": str(path_safe_root),
            "TMP": str(path_safe_root),
            "TEMP": str(path_safe_root),
        }
        temp_check = run(
            [
                sys.executable,
                "-c",
                "import os,tempfile; print(os.path.realpath(tempfile.gettempdir()), end='')",
            ],
            cwd=SOURCE_ROOT,
            environment=environment,
        )
        if not (
            temp_check.returncode == 0
            and temp_check.stderr == ""
            and temp_check.stdout == expected_root
        ):
            failures.append(
                "temporary-root selection failed: "
                f"status={temp_check.returncode!r}, stdout={temp_check.stdout!r}, "
                f"stderr={temp_check.stderr!r}, expected={expected_root!r}"
            )

        for relative_script, success_line in SMOKES.items():
            smoke_environment = dict(environment)
            if relative_script == NON_MUTATING_SMOKE:
                smoke_environment.update(
                    {
                        "GIT_DIR": str(poison_git_dir),
                        "GIT_INDEX_FILE": str(poison_index),
                        "GIT_WORK_TREE": str(poison_work_tree),
                    }
                )
            result = run(
                [sys.executable, relative_script],
                cwd=SOURCE_ROOT,
                environment=smoke_environment,
            )
            last_line = result.stdout.splitlines()[-1] if result.stdout.splitlines() else None
            if not (
                result.returncode == 0
                and result.stderr == ""
                and last_line == success_line
            ):
                failures.append(
                    f"{relative_script} failed under path-safe temporary root: "
                    f"status={result.returncode!r}, stdout={result.stdout!r}, "
                    f"stderr={result.stderr!r}"
                )
            if (
                relative_script == NON_MUTATING_SMOKE
                and tree_snapshot(poison_root) != poison_before
            ):
                failures.append(
                    f"{relative_script} mutated inherited caller-owned Git context"
                )

    if failures:
        for failure in failures:
            print(failure, file=sys.stderr)
        return 1
    print("Path-safe core consumption smoke tests passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
